Adding authentications

GilShechter opened the pull request:
### Description

This PR handles the app's authentication - login, sign up, and logout.

Close #67

### What's new

- Login and registration templates
- Relevant functions were added to the 'views.py' file
- Relevant routes were added to the 'url.py' file
- Test file
- Logout is now available through the base's header

### What it looks like

![Screenshot 2022-12-31 122908](https://user-images.githubusercontent.com/85102059/210133476-b9c45492-c677-481d-95ba-4175ccca0b3c.png)

![Screenshot 2022-12-31 122926](https://user-images.githubusercontent.com/85102059/210133478-cbcf074e-5dd6-4a78-bfa4-195387b939a7.png)

efraimvis reviewed (approved):
First of all, massive props on the human readability of the PR description - looks legit and pro, makes life just that much easier :)
Admittedly, at the moment my personal ability to thoroughly CR this PR is somewhat limited, because I haven't got much experience with the more frontend side of things in Django, but here's my main takeaway:

This is a big, big PR. Yes, you've made it clear that it relies on another PR and includes its commits at the moment, but even then - its still big. Lots of files, lots of code. Please consider trying to sift out what changes you can possible split across a number of other PRs, or omit for the moment. Please remember that there's this "PR size effect" that tends to make CR-ers in general, and our mentors in particular, less likely to feel comfortable reviewing a big and complicated PR.

GalOved reviewed (approved):
I thing you do great job.
I suggest to separate things into several different PRs
For example, all the html and css pages in a separate PR.

GalOved commented:
> First of all, massive props on the human readability of the PR description - looks legit and pro, it makes life just that much easier :)
> 
> Admittedly, at the moment my personal ability to thoroughly CR this PR is somewhat limited, because I haven't got much experience with the more frontend side of things in Django, but here's my main takeaway:
> 
> This is a big, big PR. Yes, you've made it clear that it relies on another PR and includes its commits at the moment, but even then - it's still big. Lots of files, lots of code. Please consider trying to sift out what changes you can possible split across a number of other PRs, or omit for the moment. Please remember that there's this "PR size effect" that tends to make CR-ers in general, and our mentors in particular, less likely to feel comfortable reviewing a big and complicated PR.

I think all the masiv

GilShechter commented:
I know it's quite a big PR, but it will be more complicated, I think, to split it.
Assuming that #67 will be closed, that will leave us with -
1. HTML files for 2 pages that are related to each other
2. CSS file that both HTML files share between them
3. Views functions that are also related
4. Test file, which is a must
5. The constant file to make the tests more readable
6. URL routes

I suggest we wait for the previous PR to be merged and maybe it will be easier to understand this one.

GalOved commented:
> I know it's quite a big PR, but it will be more complicated, I think, to split it. Assuming that #62 will be closed, that will leave us with -
> 
> 1. HTML files for 2 pages that are related to each other
> 2. CSS file that both HTML files share between them
> 3. Views functions that are also related
> 4. Test file, which is a must
> 5. The constant file to make the tests more readable
> 6. URL routes
> 
> I suggest we wait for the previous PR to be merged and maybe it will be easier to understand this one.

I changed my mind about that and I think you're right

svatares commented:
@GilShechter, it seems very clear that this PR could be split into three different PRs, one for login, one for registration, and one for log-out.
For the current PR, we won't ask to split it.
But please keep in mind that each functional/behavior should be pushed in a different PR for future work.

natifridman commented on tasties_app/views.py:
```diff
@@ -1,6 +1,54 @@
+            login(request, user)
+            return redirect('index')
+        else:
+            messages.info(request, 'Username OR password is incorrect')
```
Why do you need these messages? where do you use it?

GilShechter replied:
This message will pop up when you insert an invalid username or password on the login page.
It looks like this:

![incorrect](https://user-images.githubusercontent.com/85102059/210235384-610c3ce1-68e3-4d92-a24f-586f0ec79d99.png)

And it comes from these lines in `login.html`:

![Screenshot 2023-01-02 150804](https://user-images.githubusercontent.com/85102059/210235610-97f3d67e-4fb0-4edd-99f2-f38fd8c296cc.png)

natifridman replied:
Thanks

natifridman reviewed (changes requested):
You shouldn't use HTTP status code 200 for errors. You should use 30x or 40x

svatares reviewed (changes requested):
Left a couple of comments and also join Nati's comments regarding the response code in invalid cases, it shouldn't be 200.

GilShechter commented on tasties_app/tests/test_authentication.py:
```diff
@@ -0,0 +1,58 @@
+    def test_invalid_login(self, client, signed_up_credentials):
+        response = client.post('/login/', data={'username': VALID_USER, 'password': INVALID_PASSWORD_MISMATCH})
+        assert response.status_code == 200
+        assert b"Login" in response.content
```
I tried using this assertion but I got this error saying that `response` has to such an attribute:

![Screenshot 2023-01-02 145630](https://user-images.githubusercontent.com/85102059/210234366-b7ca4d12-52b8-460b-94ac-f3234e682a5e.png)

Any idea why?

natifridman replied:
I think you should add [this](https://github.com/redhat-beyond/ResuMe/blob/9d9b7369625fcb044f91cde86c85c91fdaf44ddd/ResuMe/urls.py#L25) to the urls.py

GilShechter commented on tasties_app/tests/test_authentication.py:
```diff
@@ -0,0 +1,58 @@
+
+    def test_invalid_login(self, client, signed_up_credentials):
+        response = client.post('/login/', data={'username': VALID_USER, 'password': INVALID_PASSWORD_MISMATCH})
+        assert response.status_code == 200
```
I tried to show here that I'm staying on the same page instead of being redirected to another page (302).
So if I assert `response.status_code != 302` will it be better?

natifridman replied:
There is no redirect? how the user should know that there is an error?

GilShechter replied:
No, you can see in the `register` method in `views.py` that if the form is invalid it will add an error message to `messages`, the page will be rendered, and then the error will bump. This is why the status code will be 200.

natifridman replied:
ok, thanks

GilShechter replied:
So should I use `assert response.status_code != 302` To make it more readable?

natifridman replied:
I think it's better to keep it with status code 200 and make sure it's the correct template

GilShechter commented on tasties_app/forms.py:
```diff
@@ -0,0 +1,18 @@
+
+
+class CreateUserForm(UserCreationForm):
+    password1 = forms.CharField(max_length=16, widget=forms
+                                .PasswordInput(attrs={'class': 'form-control', 'placeholder': 'Password'}))
+    password2 = forms.CharField(max_length=16, widget=forms
```
Let me explain:
I'm inheriting `UserCreationForm`, which already has those fields:
`password1` for the password
`password2` for password confirmation

So in those lines 7-9 I'm just overriding the existing fields because I want them to include the `form-control` class and placeholders.

I know it's not very pythonish, but in this case, I have to leave it as is.

You can see it in the [docs](https://docs.djangoproject.com/en/1.8/_modules/django/contrib/auth/forms/#UserCreationForm).

svatares replied:
Cool, I forgot about that. Thanks for the clarification.

GilShechter commented on tasties_app/tests/test_authentication.py:
```diff
@@ -0,0 +1,58 @@
+    def test_invalid_email_registration(self, client):
+        response = client.post('/register/', data={'username': VALID_USER, 'email': INVALID_EMAIL,
+                                                   'password1': VALID_PASSWORD, 'password2': VALID_PASSWORD})
+        assert response.status_code == 200
+        assert b"Register Account" in response.content
```
Fixed the same assertion in `test_invalid_login` but couldn't do the same here.
Is there another way of checking the URL?

natifridman replied:
What is the error you get? maybe you can check `response.url`?

GilShechter replied:
I'm getting `HttpResponse has no attribute 'url' `
And same for `template_name`.

natifridman replied:
I think you need to create a view class or use the [generic view](https://docs.djangoproject.com/en/4.1/topics/class-based-views/#usage-in-your-urlconf)

GilShechter replied:
At last, find a way of checking the URL.

natifridman replied:
Great!

GilShechter commented:
I rebased this PR to include the commits from #62.
Also made changes due to your comments:
- I removed the `login_user` method and now using the same method from `django.contrib.auth`, so it also caused changes in the settings file (Added `LOGIN_REDIRECT_URL` which was required) and changed the URL and the tests.
- In the registration template, changes the email-form's placeholder as suggested by @svatares.

svatares commented on tasties_app/templates/tasties_app/base.html:
```diff
@@ -33,27 +33,32 @@
+      <a class="navbar-brand" style="color: #10654e" href="{% url 'index' %}">
         <img
           src="{% static 'images/logo.png' %}"
+          width="30"
+          height="25"
           class="d-inline-block align-top"
           alt="tasties-logo"
+          style="margin-top: 5px"
```
Didn't we agree in the previous PR you had, that all of those configurations are going to be in the CSS file?

GilShechter replied:
Yes, I made a mistake while resolving conflicts with this file.
Thanks for noticing.

svatares reviewed (changes requested):
please see my comment on the base.html file

GalOved reviewed (dismissed):
Looking good to me

GilShechter commented:
> I rebased this PR to include the commits from #62. Also made changes due to your comments:
> 
> * I removed the `login_user` method and now using the same method from `django.contrib.auth`, so it also caused changes in the settings file (Added `LOGIN_REDIRECT_URL` which was required) and changed the URL and the tests.
> * In the registration template, changes the email-form's placeholder as suggested by @svatares.

I noticed that due to my changes here in the login view method, the login form won't show any messages.
So don't merge yet.

GilShechter commented:
> > I rebased this PR to include the commits from #62. Also made changes due to your comments:
> > 
> > * I removed the `login_user` method and now using the same method from `django.contrib.auth`, so it also caused changes in the settings file (Added `LOGIN_REDIRECT_URL` which was required) and changed the URL and the tests.
> > * In the registration template, changes the email-form's placeholder as suggested by @svatares.
> 
> I noticed that due to my changes here in the login view method, the login form won't show any messages. So don't merge yet.

Fixed.
I brought back the first method that I wrote for the login view,
and used the solution that I found for testing the URL.

natifridman reviewed (dismissed):
Looks good, nice work!

GilShechter commented:
Squshed commits.

GilShechter commented on tasties_app/templates/tasties_app/base.html:
```diff
@@ -72,7 +72,7 @@
             <div class="dropdown-menu" aria-labelledby="dropdownMenuLink">
               <a class="dropdown-item" href="#">Add recipe</a>
-              <a class="dropdown-item" href="#">Logout</a>
+              <a class="dropdown-item" href="{% url 'logout' %}">Logout</a>
```
The default `/` will redirect unauthorized users to the login page,
and then when the user login he can see the `logout`

natifridman replied:
So unauthorized users won't be able to browse recipes?

GilShechter replied:
No, according to the DFD file, only users can.

natifridman replied:
Great, thanks
